### ISSUE TITLE
Incorrect warning regarding no matching class member found

### ISSUE THREAD

**Issue by albert-github** [doxygen/doxygen#7900]
**Describe the bug**
Incorrect warning regarding no matching class member found, regression (see Version and Additional context).
With the example as shown in the attachment we get the warning:

> .../FXGLPoint.cpp:8: warning: no matching class member found for
>   FXGLPoint::FXGLPoint(float x, FXfloat y, FXfloat z)
> Possible candidates:
>   'FXGLPoint::FXGLPoint(FXfloat x, FXfloat y, FXfloat z)'
>   'FXGLPoint::FXGLPoint()'

though the definition and declaration are in both: `FXGLPoint::FXGLPoint(FXfloat x, FXfloat y, FXfloat z)`
note also `float` instead of `FXfloat` as shown the second line of the warning.

The problem is a bit specific for doxygen as doxygen does handle all files separately and thus the handling of `FXGLPoint.cpp` has all definitions but the handling of `FXGLPoint.h` hasn't, the compiler compensates normally for this as there should always be a `#include "fx.h"` present.

The problem was found during the investigation of the warnings in the "fox  1.7.71" project as reported by Fossies.

**Expected behavior**
No warning and correct links

**To Reproduce**
Example: [example.tar.gz](https://github.com/doxygen/doxygen/files/4897000/example.tar.gz)

**Version**
1.8.19 (cbf83033232f8c76836bd8f086bfad146ed74e00). Problem was not present in the 1.8.18 release and version 1.8.19 (cd581388f3d013c501e3cefbaf3e81cf93d46fcb) from April 19, 2020.
The problem surfaced with the next commit version: 1.8.19 (9e2b6a2dac870b5daabde118c5327281e3f79ba2), from April 24, 2020 ("Performance improvements after profiling").

Note: the same message appeared in versions till 1.8.16 but was gone in 1.8.17.

**Additional context**
Most likely candidate: in `util.cpp` the function `matchArguments2`

The idea was to add:
```
  if ((srcScope && dstScope && srcScope->name()==dstScope->name() && srcA.type==dstA.type)  ||
      (!srcScope && !dstScope && srcA.type==dstA.type))
  {
    MATCH
    return TRUE;
  }
```
after the resetting of the `canType` in `matchArguments2` but this lead to some incorrect links in the mentioned fox project and missing struct definitions (and possible other missing information) in the CGAL project.

**Comment by doxygen** [comment]
Doxygen tries to resolve each type to its canonical form before matching it. In your example there is no type information available in `FXGLPoint.h` so the canonical type for `FXfloat` is simply `FXfloat`. In the source file due to the `#include` the same `FXfloat` type can be resolved to a `float` type. When then comparing the declaration in the header against the definition no match is found. 

The common practice and assumption made by doxygen is that the header file is self contained, i.e. it includes the types it uses. This doesn't hold in your example, so I understand why doxygen has a problem with it. The only thing that is not clear yet is why it did work before.

**Comment by albert-github** [comment]
It is indeed strange that it worked in the past.

Furthermore it is a bit inconsistent that we see:
```
FXGLPoint::FXGLPoint(float x, FXfloat y, FXfloat z)
```
and not:
```
FXGLPoint::FXGLPoint(float x, float y, float z)
```
1 times 'float' against 3 times; but this is due to the way doxygen tries to resolve the arguments and as soon as a "miss" is found it stops resolving the rest of the arguments.

**Comment by doxygen** [comment]
Doxygen has a fallback to check if there is a literal match between the argument lists before giving up, that triggered in other releases but was (now) wrongly using the canonical type.

**Comment by albert-github** [comment]
Just tested against "fox 1.7.71" no new warnings, number of warnings is reduced slightly.
Just tested against CGAL and no new warnings.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.19.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
